src/reconcile/rec_visibility.c

```diff
@@ -637,30 +637,66 @@ __rec_calc_upd_memsize(WT_UPDATE *onpage_upd, WT_UPDATE *tombstone, size_t upd_m
 
 /*
  * __rec_upd_select --
- *     Select the update to write to disk image.
+ *     Select the update to write to disk image. @param write_prepare True if we should write the
+ *     update as a prepared update (prepare timestamp is stable but durable timestamp is not), false
+ *     to write as committed.
  */
 static int
 __rec_upd_select(WT_SESSION_IMPL *session, WTI_RECONCILE *r, WT_UPDATE *first_upd,
   WTI_UPDATE_SELECT *upd_select, WT_UPDATE **first_txn_updp, bool *has_newer_updatesp,
-  size_t *upd_memsizep)
+  bool *write_prepare, size_t *upd_memsizep)
 {
     WT_CONNECTION_IMPL *conn;
-    WT_UPDATE *upd;
+    WT_UPDATE *upd, *prepare_rollback_tombstone;
     wt_timestamp_t max_ts;
     uint64_t max_txn, session_txnid, txnid;
     uint8_t prepare_state;
     bool is_hs_page, seen_prepare;
 
     conn = S2C(session);
+    prepare_rollback_tombstone = NULL;
     max_ts = WT_TS_NONE;
     max_txn = WT_TXN_NONE;
     is_hs_page = F_ISSET(session->dhandle, WT_DHANDLE_HS);
     session_txnid = __wt_atomic_loadv64(&WT_SESSION_TXN_SHARED(session)->id);
     seen_prepare = false;
+    *write_prepare = false;
 
     for (upd = first_upd; upd != NULL; upd = upd->next) {
-        if ((txnid = upd->txnid) == WT_TXN_ABORTED)
-            continue;
+        *write_prepare = false;
+        WT_ACQUIRE_READ(txnid, upd->txnid);
+        /*
+         * If we have seen a globally visible tombstone that rolled back a prepared update, we must
+         * now see an aborted prepared update.
+         */
+        WT_ASSERT(session,
+          prepare_rollback_tombstone == NULL ||
+            (txnid == WT_TXN_ABORTED && upd->prepare_state == WT_PREPARE_INPROGRESS));
+        if (txnid == WT_TXN_ABORTED) {
+            if (!F_ISSET(conn, WT_CONN_PRESERVE_PREPARED))
+                continue;
+
+            WT_READ_ONCE(prepare_state, upd->prepare_state);
+            if (prepare_state != WT_PREPARE_INPROGRESS)
+                continue;
+
+            /* Ignore the prepared update if the rollback timestamp is stable. */
+            if (upd->upd_rollback_ts <= r->rec_start_pinned_stable_ts) {
+                /* WT_ASSERT(session, upd->upd_rollback_ts != WT_TS_NONE); */
```

Comment:
I understand why this can fire but I think it poses a data corruption risk. If their is no rollback ts, we should treat the rollback ts as not stable. So we should do:

if (upd->upd_rollback_ts != WT_TS_NONE && upd->upd_rollback_ts <= r->rec_start_pinned_stable_ts)